File: javascript/007/007.js
"use strict";

const sieveOfAtkin = function(limit) {
  let primes = [];

  let sieve = [];

  for (let x = 0; x < limit; x += 1) {
    sieve.push(false);
  }

  const squareRootOfLimit = Math.round(Math.sqrt(limit));

  let n = 0;

  for (let x = 1; x <= squareRootOfLimit; x += 1) {
    for (let y = 1; y <= squareRootOfLimit; y += 1) {
      n = (4 * x * x) + (y * y);

      if (n <= limit && (n % 12 === 1 || n % 12 === 5)) {
        sieve[n] = !sieve[n];
      }

      n = (3 * x * x) + (y * y);

      if (n <= limit && n % 12 === 7) {
        sieve[n] = !sieve[n];
      }

      n = (3 * x * x) - (y * y);

      if (x > y && n <= limit && n % 12 === 11) {
        sieve[n] = !sieve[n];
      }
    }
  }

  sieve[2] = true;

  sieve[3] = true;

  primes.push(2);

  primes.push(3);

  n = 5;

  while (n <= squareRootOfLimit) {
    if (sieve[n]) {
      let i = n * n;

      while (i < limit) {
        sieve[i] = false;

        i += n * n;
      }

      primes.push(n);
    }

    n += 2;
  }

  while (n < limit) {
    if (sieve[n]) {
      primes.push(n);
    }

    n += 2;
  }

  return primes;
};

const n = 10001;

console.log(sieveOfAtkin(Math.round(((n * Math.log(n)) + (n * (Math.log(Math.log(n)))))))[n - 1]);

File: csharp/007/007.cs
using System;
using System.Collections.Generic;

public class Euler {
    static public void Main() {
        ulong n = 10001;

        ulong limit = (ulong)((n * Math.Log(n)) + (n * (Math.Log(Math.Log(n)))));

        Console.WriteLine("{0}", SieveOfAtkin(limit)[(int)n - 1]);
    }

    static public List<ulong> SieveOfAtkin(ulong limit) {
        List<ulong> primes = new List<ulong>();

        List<bool> sieve = new List<bool>();

        for (ulong i = 0; i < limit + 1; i++) {
            sieve.Add(false);
        }

        ulong squareRootOfLimit = (ulong)Math.Sqrt(limit);

        ulong n;

        for (ulong x = 1; x <= squareRootOfLimit; x++) {
            for (ulong y = 1; y <= squareRootOfLimit; y++) {
                n = (4 * x * x) + (y * y);

                if (n <= limit && (n % 12 == 1 || n % 12 == 5)) {
                    sieve[(int)n] = !sieve[(int)n];
                }

                n = (3 * x * x) + (y * y);

                if (n <= limit && n % 12 == 7) {
                    sieve[(int)n] = !sieve[(int)n];
                }

                n = (3 * x * x) - (y * y);

                if (x > y && n <= limit && n % 12 == 11) {
                    sieve[(int)n] = !sieve[(int)n];
                }
            }
        }

        sieve[2] = true;

        sieve[3] = true;

        primes.Add(2);

        primes.Add(3);

        n = 5;

        while (n <= squareRootOfLimit) {
            if (sieve[(int)n]) {
                ulong i = n * n;

                while (i < limit) {
                    sieve[(int)i] = false;

                    i += n * n;
                }

                primes.Add(n);
            }

            n += 2;
        }

        while (n < limit) {
            if (sieve[(int)n]) {
                primes.Add(n);
            }

            n += 2;
        }

        return primes;
    }
}
